Contribute guidelines: amend documentation for policies around modifying others' work (#26046)
zephyrproject-rtos/zephyr: doc/contribute/modifying_contributions.rst

.. _modifying_contributions:

Modifying Contributions made by other developers
************************************************

Scenarios
#########

In general, Zephyr contributors and collaborators are encouraged to assist
as reviewers in pull requests, so that patches may be approved and merged
to Zephyr's main branch as part of the original pull requests. The authors
of the pull requests are responsible for amending their original commits
following the review process.

There are occasions, however, when a contributor might need to modify patches
included in pull requests that are submitted by other Zephyr contributors.

> For instance, this is the case when

Reviewer: Generally paragraphs that precede a bullet list in rst text in Zephyr end with either `:` or `::`.  Having neither doesn't seem right.

Author: Agreed. Fixed this now

  pull requests in order to

  * integrate useful content which is part of a stale pull request, or
  * get content merged to the project's main branch as part of a larger
    patch

* a developer pushes onto a branch or pull request opened by another
  contributor in order to

  * assist in revising pull requests and get the patches merged
    to the project's main branch
  * finalize review in stale pull requests

Recommended policies
####################

A developer who intends to picking and modifying patches sent by
another contributor should

* clarify in their pull request the reason for doing that, instead
  of assisisting in getting the patches merged in their original
  pull request, and
* invite the original author of the patches in their pull request review.

A developer who intends to force-push onto a branch or pull request of
another Zephyr contributor should clarify in the pull request the reason
for pushing and for modifying the existing patches (e.g. stating that it
is done for assisting in the pull request review when the pull
request author is not able to do so).

.. note::
  Developers should try to limit the above practice to pull requests identified

>   as *stale*.

Reviewer: The criteria and process for characterizing a PR as *stale* do not appear to be defined.

Also I sometimes do this on non-stale PRs when then original author agrees to let me, generally because of differential git skills and the complexity of getting the exact desired solution in place with a clean commit history.

Author: Agreed, this needs to be defined somewhere.

Author: @pabigot stale PRs and ISsues are defined here:
https://docs.zephyrproject.org/latest/development_process/dev_env_and_tools.html

I will add a link.

Author: > Also I sometimes do this on non-stale PRs when then original author agrees to let me, generally because of differential git skills and the complexity of getting the exact desired solution in place with a clean commit history.

Agreed. I believe, the way I wrote the sentence still allows you to continue doing this:
> should **try to limit** the above practice

But it is not encouraged as a long-term solution. Agreeing with original author is something we refer to later (for modifying content) but in theory it could also apply here, so I think we are sort of covered.

If the original patches are substantially modified, the developer could
request the original author to acknowledge that the modified patches may
be merged while having the original sign-off line and author identity.
If the original author is not responsive the developer should, alternatively,
be able to submit the modified patches as their *own* work (i.e. with their
*own* sign-off line and author identity). In the latter case, the developer

> may want to indentify in the commit message(s) the original source the

Reviewer: s/indentify/identify/

The criteria and process for characterizing an original author as *unresponsive* are not defined.

Also I would propose that if any material from the original is retained the signed-off-by line of the original should be kept, based on [this discussion](https://stackoverflow.com/a/1962112/236848) of the intent behind that statement (chain-of-responsibility).

Author: 1. Typo is now fixed.
2. unresponsive authors: yeap, this needs to be defined explicitly, but it is not part of this change, so I prefer to address it outside this PR

>Also I would propose that if any material from the original is retained the signed-off-by line of the original should be kept, based on this discussion of the intent behind that statement (chain-of-responsibility).

We discussed this extensively in the Process Forum (participation is open to all, btw). Retaining the original sign-off cannot be required in all cases; this is exactly the problem we are trying to solve: picking and heavily modifying commits of others, while retaining original sign-off. We don't want to do that, unless the original authors ACKed this.

  * get content merged to the project's main branch as part of a larger
    patch

* a developer pushes onto a branch or pull request opened by another
  contributor in order to

  * assist in revising pull requests and get the patches merged
    to the project's main branch
  * finalize review in stale pull requests

Recommended policies
####################

A developer who intends to picking and modifying patches sent by
another contributor should

* clarify in their pull request the reason for doing that, instead
  of assisisting in getting the patches merged in their original
  pull request, and
* invite the original author of the patches in their pull request review.

A developer who intends to force-push onto a branch or pull request of
another Zephyr contributor should clarify in the pull request the reason
for pushing and for modifying the existing patches (e.g. stating that it
is done for assisting in the pull request review when the pull
request author is not able to do so).

.. note::
  Developers should try to limit the above practice to pull requests identified
  as *stale*.

If the original patches are substantially modified, the developer could either

> * reach out to the original author and request them to acknowledge that the

Author: Simplified (based on @galak 's fb)

* clarify in their pull request the reason for doing that, instead
  of assisisting in getting the patches merged in their original
  pull request, and
* invite the original author of the patches in their pull request review.

A developer who intends to force-push onto a branch or pull request of
another Zephyr contributor should clarify in the pull request the reason
for pushing and for modifying the existing patches (e.g. stating that it
is done for assisting in the pull request review when the pull
request author is not able to do so).

.. note::
  Developers should try to limit the above practice to pull requests identified
  as *stale*.

If the original patches are substantially modified, the developer could either

* reach out to the original author and request them to acknowledge that the
  modified patches may be merged while having the original sign-off line and
  author identity, or
* submit the modified patches as their *own* work (i.e. with their
  *own* sign-off line and author identity). In this case, the developer
  may want to indentify in the commit message(s) the original source the
  submitted work is based on (mentioning, for example, the original PR

>   number).

Author: Adding a comment about the PR number as an example of how an author can link to someone else's work

.. _modifying_contributions:

Modifying Contributions made by other developers
************************************************

Scenarios
#########

> In general, Zephyr contributors and collaborators are encouraged to assist

Reviewer 2: Do you need "In general,"?

Author: Well, no. Thanks, I am removing this.

as reviewers in pull requests, so that patches may be approved and merged
to Zephyr's main branch as part of the original pull requests. The authors
of the pull requests are responsible for amending their original commits
following the review process.

There are occasions, however, when a contributor might need to modify patches
included in pull requests that are submitted by other Zephyr contributors.
For instance, this is the case when:

> * a developer picks commits submitted by other contributors into their own

Reviewer 3: cherry-picks

Author: yeap, thanks

  pull requests in order to:

  * integrate useful content which is part of a stale pull request, or
  * get content merged to the project's main branch as part of a larger
    patch

> * a developer pushes onto a branch or pull request opened by another

Reviewer 3: ```suggestion
* a developer pushes to a branch or pull request opened by another
```

Author: done, thanks

  contributor in order to:

>   * assist in revising pull requests and get the patches merged

Reviewer 3: ```suggestion
  * assist in updating pull requests in order to get the patches merged
```

Author: thanks, done.

    to the project's main branch

>   * finalize review in stale pull requests

Reviewer 3: ```suggestion
  * drive stale pull requests to completion so they can be merged
```

Author: yeap, that's fine , too. thanks

Accepted policies
#################

> A developer who intends to picking and modifying patches sent by

Reviewer 3: ```suggestion
A developer who intends to cherry-pick and potentially modify patches sent by
```

Author: done, too.

another contributor shall:

* clarify in their pull request the reason for doing that, instead
  of assisisting in getting the patches merged in their original

>   pull request, and

Reviewer 2: why instead of?

Author: Hm, I guess "instead of", here, is used to distinguish between "doing that" and "assisting in getting..." 
As it seems not to be clear, I'll replace "doing that" with the actual verb action :)

Author: fixed now

> * invite the original author of the patches in their pull request review.

Reviewer 3: ```suggestion
* invite the original author of the patches to their pull request review.
```

Author: fixed

> A developer who intends to force-push onto a branch or pull request of

Reviewer 3: ```suggestion
A developer who intends to force-push to a branch or pull request of
```

Author: fixed

> another Zephyr contributor shall clarify in the pull request the reason

Reviewer 4: Should we state that "Zephyr contributor shall clarify" _before_ doing the force-push? I.e. a user who wants to force push should first notify the author of the original PR of their intent? Just to avoid a situation when someone, suddenly, makes a force push while the original author has already made some changes localy but didn't push their work to the github repo. I assume that's the expected behavior but at the moment it is not explicit.

Author: @mnkp thanks for your comment: I would think this is rather intuitive; we should always be careful with force-pushing, that's why I think this might be not needed here. Besides:
- we stress that force-pushing to other's branch/PR should be limited to stale PRs, i.e. with un-responsive authors, and for reasons like to drive the PR to completion (when the author is not able to do themselves), so I would say the "race-condition" is not the use-case here :)

Reviewer: "Race-condition" is not the point.  The point is who has ownership of the *branch* and so should be in control of what's in it.  The branch is in the submitter's repository.  Anybody pushing to a branch in my repos without getting a positive acknowledgement from me that it's OK is not going to be well thought of.  People may manage their branches in ways we don't understand, and a force push from other content might break that management.

My personal policy is in line with that: I'll only push if the owner of the repo authorizes it.  In any other case where I was taking over a stale PR under the conditions authorized in this policy I would issue a new PR from my repo, referencing the original which would be closed.

Author: > My personal policy is in line with that: I'll only push if the owner of the repo authorizes it. In any other case where I was taking over a stale PR under the conditions authorized in this policy I would issue a new PR from my repo, referencing the original which would be closed.

@pabigot although I am fully supporting more transparency and more communication in the Project (and your proposal is ideal with respect to transparency/communication), I am afraid, this may introduce more problems than what it solves.
- asking for authorization to force push: then we need to define how to do it: Slack? E-mail? GH comment? And for how long we should wait until we flag the author as non-responsive? 
- issuing a new PR from your repo is problematic: you lose comment/review history, and will need to ask around, once more, for all the approvals you had in the original PR. IMHO, this slows down a lot the process.

I assume a common case of a PR that is already reviewed, approved, but with Git conflicts, or broken CI, and the author being on holiday for a while; a trivial force-push should do it (with proper commenting, of course, as we explain here) and for the Release Engineer it would be much easier to get the PR merged, rather than looking at a brand-new PR with the same changes.

If anyone force-pushes to your branch to get CI/conflicts fixed and get it merged; I don't clearly see why this should, in general not be well thought of.

In any case, authors can always uncheck the relevant GH box to disallow adding / modifying commits

for pushing and for modifying the existing patches (e.g. stating that it

> is done for assisting in the pull request review when the pull

Reviewer 3: ```suggestion
is done to drive the pull request review  to completion when the pull
```

Author: thanks, done

request author is not able to do so).

.. note::
  Developers should try to limit the above practice to pull requests identified
  as *stale*. Read about how to identify pull requests as stale in
  :ref:`development processes and tools <dev-environment-and-tools>`

If the original patches are substantially modified, the developer can either:

* (preferably) reach out to the original author and request them to
  acknowledge that the modified patches may be merged while having
  the original sign-off line and author identity, or
* submit the modified patches as their *own* work (i.e. with their
  *own* sign-off line and author identity). In this case, the developer

>   may want to identify in the commit message(s) the original source the

Reviewer 3: ```suggestion
  shall identify in the commit message(s) the original source the
```
I think we should always mention where the original work comes from

Author: @dleach02 @carlescufi no objections to this.

  submitted work is based on (mentioning, for example, the original PR
  number).

.. note::

>   Contributors should be un-checking the box *“Allow Edits By Maintainers"*

Reviewer 3: ```suggestion
  Contributors should be uncheck the box *“Allow Edits By Maintainers"*
```

Author: thanks. I suppose it should be "should uncheck"
